src/stan/services/pathfinder/single.hpp

```diff
@@ -459,19 +460,18 @@ inline taylor_approx_t taylor_approximation(
  * matrix of samples, and an unsigned integer for number of times the log prob
  * functions was called
  */
-template <bool ReturnLpSamples, typename EigMat, typename EigVec,
-          std::enable_if_t<ReturnLpSamples>* = nullptr>
-inline auto ret_pathfinder(int return_code, EigVec&& lp_ratio, EigMat&& samples,
-                           const std::atomic<size_t>& lp_calls) {
-  return std::make_tuple(return_code, std::forward<EigVec>(lp_ratio),
-                         std::forward<EigMat>(samples), lp_calls.load());
-}
-
-template <bool ReturnLpSamples, typename EigMat, typename EigVec,
-          std::enable_if_t<!ReturnLpSamples>* = nullptr>
-inline auto ret_pathfinder(int return_code, EigVec&& lp_ratio, EigMat&& samples,
-                           const std::atomic<size_t>& lp_calls) noexcept {
-  return return_code;
+template <bool ReturnLpSamples, typename EigVec, typename ParamWriter>
+inline auto ret_pathfinder(int return_code, EigVec&& elbo_est,
+                           const std::atomic<size_t>& lp_calls,
+                           ParamWriter&& /* params */) noexcept {
+  if constexpr (ReturnLpSamples) {
+    return std::make_tuple(return_code, std::forward<EigVec>(elbo_est),
+                           lp_calls.load());
+  } else if constexpr (stan::callbacks::is_multi_writer_v<ParamWriter>) {
+    return std::pair(return_code, lp_calls.load());
```

Comment:
The docstring of this function needs to be updated with the new template parameters and I think this needs to be explained -- is there a reason to not give this a name as another bool tparam?

Reply:
I think I need to clean this up. ReturnLpSamples will only be true if we are not doing psis.